url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+     * とすると、イミディエート imm32 は下記のように求められます。
+     *
+     * imm32 = rotateRight(immed_8, rotate_imm * 2)
+     *
+     * @param inst ARM 命令
+     * @return イミディエート
+     */
+    public int getAddrMode1Imm(InstructionARM inst) {
+        int rotR = inst.getField(8, 4);
+        int imm8 = inst.getField(0, 8);
+
+        return Integer.rotateRight(imm8, rotR * 2);
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランド、
+     * イミディエートシフトを取得します。
+     *
+     * 条件:
+     * I ビット（ビット[25]）: 0
+     * ビット[4]: 0
+     *
+     * 該当するオペランド:
+     * 数値はビット[6:4] の値を示す。
+     * 0b000: データ処理オペランド - レジスタ
+     * 0b000: データ処理オペランド - イミディエート論理左シフト
+     * 0b010: データ処理オペランド - イミディエート論理右シフト
+     * 0b100: データ処理オペランド - イミディエート算術右シフト
+     * 0b110: データ処理オペランド - イミディエート右ローテート
+     * 0b110: データ処理オペランド - 拡張付き右ローテート
+     *
+     * @param inst ARM 命令
+     * @return イミディエートシフトオペランド
+     */
+    public int getAddrMode1ImmShift(InstructionARM inst) {
+        int shift_imm = inst.getField(7, 5);
+        int shift = inst.getField(5, 2);
+        int rm = inst.getRmField();
+
+        switch (shift) {
+        case 0:
+            if (shift_imm == 0) {
+                //レジスタ
+                return getReg(rm);
+            } else {
+                //イミディエート論理左シフト
+                return getReg(rm) << shift_imm;
+            }
+        case 1:
+            //イミディエート論理右シフト
+            if (shift_imm == 0) {
+                return 0;
+            } else {
+                return getReg(rm) >>> shift_imm;
+            }
+        case 2:
+            //イミディエート算術右シフト
+            if (shift_imm == 0) {
+                if (BitOp.getBit32(getReg(rm), 31)) {
+                    return 0xffffffff;
+                } else {
+                    return 0;
+                }
+            } else {
+                return getReg(rm) >> shift_imm;
+            }
+        case 3:
+            if (shift_imm == 0) {
+                //拡張付き右ローテート
+                if (getCPSR().getCBit()) {
+                    return (1 << 31) | (getReg(rm) >>> 1);
+                } else {
+                    return getReg(rm) >>> 1;
+                }
+            } else {
+                //イミディエート右ローテート
+                return Integer.rotateRight(getReg(rm), shift_imm);
+            }
+        default:
+            //do nothing
+            break;
+        }
+
+        throw new IllegalArgumentException("Unknown Imm Shift " +
+                String.format("0x%08x, shift:%d.",
+                        inst.getInst(), shift));
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランド、
+     * レジスタシフトを取得します。
+     *
+     * 条件:
+     * I ビット（ビット[25]）: 0
+     * ビット[4]: 1
+     *
+     * 該当するオペランド:
+     * 数値はビット[6:4] の値を示す。
+     * 0b001: データ処理オペランド - レジスタ論理左シフト
+     * 0b011: データ処理オペランド - レジスタ論理右シフト
+     * 0b101: データ処理オペランド - レジスタ算術右シフト
+     * 0b111: データ処理オペランド - レジスタ右ローテート
+     *
+     * @param inst ARM 命令
+     * @return レジスタシフトオペランド
+     */
+    public int getAddrMode1RegShift(InstructionARM inst) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.